src/components/timeline/TrackPowerGraph.js

```diff
@@ -0,0 +1,458 @@
+  _onMouseMove = (event: SyntheticMouseEvent<HTMLDivElement>) => {
+    const { pageX: mouseX, pageY: mouseY } = event;
+    // Get the offset from here, and apply it to the time lookup.
+    const { left } = event.currentTarget.getBoundingClientRect();
+    const { width, rangeStart, rangeEnd, counter, interval } = this.props;
+    const rangeLength = rangeEnd - rangeStart;
+    const timeAtMouse = rangeStart + ((mouseX - left) / width) * rangeLength;
+
+    if (counter.sampleGroups.length === 0) {
+      // Gecko failed to capture samples for some reason and it shouldn't happen for
+      // malloc counter. Print an error and bail out early.
+      throw new Error('No sample group found for power counter');
+    }
+    const { samples } = counter.sampleGroups[0];
+
+    if (
+      timeAtMouse < samples.time[0] ||
+      timeAtMouse > samples.time[samples.length - 1] + interval
+    ) {
+      // We are outside the range of the samples, do not display hover information.
+      this.setState({ hoveredCounter: null });
+    } else {
+      // When the mouse pointer hovers between two points, select the point that's closer.
+      let hoveredCounter;
+      const bisectionCounter = bisectionRight(samples.time, timeAtMouse);
+      if (bisectionCounter > 0 && bisectionCounter < samples.time.length) {
+        const leftDistance = timeAtMouse - samples.time[bisectionCounter - 1];
+        const rightDistance = samples.time[bisectionCounter] - timeAtMouse;
+        if (leftDistance < rightDistance) {
+          // Left point is closer
+          hoveredCounter = bisectionCounter - 1;
+        } else {
+          // Right point is closer
+          hoveredCounter = bisectionCounter;
+        }
+      } else {
+        hoveredCounter = bisectionCounter;
+      }
+
+      if (hoveredCounter === samples.length) {
+        // When hovering the last sample, it's possible the mouse is past the time.
+        // In this case, hover over the last sample. This happens because of the
+        // ` + interval` line in the `if` condition above.
+        hoveredCounter = samples.time.length - 1;
+      }
+
+      this.setState({
+        mouseX,
+        mouseY,
+        hoveredCounter,
+      });
+    }
+  };
+
+  _renderTooltip(counterIndex: number): React.Node {
+    const { counter, interval } = this.props;
+    const samples = counter.sampleGroups[0].samples;
+    if (samples.length === 0) {
+      // Gecko failed to capture samples for some reason and it shouldn't happen for
+      // malloc counter. Print an error and bail out early.
+      throw new Error('No sample found for power counter');
+    }
+    const powerUsageInPwh = samples.count[counterIndex]; // picowatt-hour
+    const sampleTimeDeltaInMs =
+      counterIndex === 0
+        ? interval
+        : samples.time[counterIndex] - samples.time[counterIndex - 1];
+    const power =
+      ((powerUsageInPwh * 1e-12) /* pWh->Wh */ / sampleTimeDeltaInMs) *
+      1000 * // ms->s
+      3600; // s->h
```

Comment:
I was trying to think about not losing precision here. But I feel like that what you wrote is the best we can do.